### PR TITLE
SDK-1486: Python update docs

### PR DESCRIPTION
**Pull request by dandavison**
[[rendered](https://temporal-documentation-aj72f1nyt.preview.thundergun.io/develop/python/message-passing)]

This is a new version of the message-passing documentation for Python. It follows on from the new top-level section on https://docs.temporal.io/encyclopedia/workflow-message-passing that @drewhoskins-temporal recently wrote. The significant changes here are:

- There is now a code sample at the top of the page that will immediately give many visitors to the page what they're looking for: it is minimal, but it shows how to define Query, Signal, and Update handlers, and an Update validator, and has one-line comments introducing the concepts.

- Following the style taken in https://docs.temporal.io/encyclopedia/workflow-message-passing, the document recognizes that many aspects of Signal and Update handling can be covered together. This avoids a lot of repetition.

- In particular, a new section named `Message handler patterns` gives Python-specific discussion and examples of the content outlined in [encyclopedia/workflow-message-passing](https://docs.temporal.io/encyclopedia/workflow-message-passing).

- Similarly, there was no need to discuss dynamic handlers separately for Signal, Query, and Update since the signatures are the same.

The structure of the document is (I'll keep this up-to-date it it changes)

```
## Writing message handlers
  ### Query handlers {#handle-query}
  ### Signal handlers {#handle-signal}
  ### Update handlers and validators {#handle-update}

## Sending messages
  ### Sending a Query {#send-query}
  ### Sending a Signal
    #### Sending a Signal from a Client {#send-signal-from-client}
    #### Sending a Signal from a Workflow {#send-signal-from-workflow}
    #### Signal-With-Start {#signal-with-start}
  ### Sending an Update {#send-update}
    #### 1. Use [`execute_update`](https://python.temporal.io/temporalio.client.WorkflowHandle.html#execute_update) to wait for the update to complete
    #### 2. Use [`start_update`](https://python.temporal.io/temporalio.client.WorkflowHandle.html#start_update) to receive a handle as soon as the update is accepted or rejected
    #### Non-type safe APIs

## Message handler patterns {#message-handler-patterns}
  ### Async handlers {#async-handlers}
    #### Waiting
    #### Use `asyncio.Lock` to prevent concurrent handler execution {#control-handler-concurrency}
    #### Finishing handlers before the Workflow completes {#wait-for-message-handlers}

## Troubleshooting
  ### Problems when sending a Signal
  ### Problems when sending an Update
  ### Problems when sending a Query
## Dynamic Handlers {#dynamic-handler}
  ### Set a Dynamic Signal, Query, or Update handler {#set-a-dynamic-signal}
  ### Set a Dynamic Workflow {#set-a-dynamic-workflow}
  ### Set a Dynamic Activity {#set-a-dynamic-activity}
```

### TODO:
- [x] `get_update_info()`
- [x] add "workflow execution already completed" code

### CODE REVIEW

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [inline docs/develop/python/message-passing.mdx:347]: Alternatively perhaps we want the handler in the example code to be `async`, although making that be necessary would complicate the code.

File: docs/develop/python/message-passing.mdx
Comment (drewhoskins-temporal) [inline docs/develop/python/message-passing.mdx:387]: Plz include an uplink in each of these sections, and consider whether to reduce the amount of prose instead.

**Review by drewhoskins-temporal — COMMENTED** [review]
Looks good, mostly minor comments.  Main question is just to refine our posture on how much explanatory text goes here vs in the overview docs, and how we link to the overview docs, which I've flagged in a scattershot fashion in the inline comments.
It's convenient to have things here for people who like to dive right in.  But I'm worried about a middle ground where we say some important things but not others.
I've had two experiences with temporal's language specific docs where they didn't answer my question, and I wasn't aware that the answer was in the overview docs.  (that is, wasn't aware that there were overview docs on those topics and at this time I had not internalized the overall structure of the docs.)
So, liberal upward linking is probably the main thing I would recommend, and then use judgment about whether to repeat things here (and maybe overall I would recommend doing it less than you are and rely more on the links.  This would ensure people are getting the best version of our intro on these topics, which we can maintain carefully.).

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:31]: Right, I did consider that. However, the directory would need a rename since my sample very much covers Query also. Do you think renaming your directory, e.g. to `message_passing`, might make sense? I don't have a strong opinion there -- I can also see that having a place to discuss async handlers and the pitfalls thereof has value.

No particular love for `hello/`, but I'm going to push for maintaining single-file structure for simple toy samples like this, so that they can simply be run by `python the_example.py` from within a virtualenv.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:78]: Sure! Yeah, I was very focused on keeping sentence length down in those comments, since we want this code sample to appear "light weight".

Changing to

> A Query handler returns a value: it can inspect but must not mutate the Workflow state.

(Unfortunately, it _can_ mutate, so I prefer "must not")

File: docs/develop/python/message-passing.mdx
Comment (cretz) [inline docs/develop/python/message-passing.mdx:374]: This is such an important concept and `wait_condition` is such a powerful/common tool for a workflow, it may also deserve its own section unrelated to message. Doesn't affect this PR, but we often find people don't know that this tool exists.

**Review by cretz — APPROVED** [review]
Some nits here and there, but overall LGTM

**Review by dandavison — COMMENTED** [review]
Thanks for the review @drewhoskins-temporal! I believe I've responded to all your comments. I've added several more uplinks.

Turning to @cretz's review now which also looks very helpful.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:221]: Good call, added this to the `Waiting` section:

> This is necessary if your Signal or Update handlers require something in the main Workflow method to be done first, since a Signal or Update handler can execute before the main Workflow method has started.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks, changed to

> : that means that the Worker must not be offline, and must not respond too slowly

**Review by dandavison — COMMENTED** [review]
Dealt with first batch of @cretz's comments -- thanks!

File: docs/develop/python/message-passing.mdx
Comment (cretz) [reply docs/develop/python/message-passing.mdx:33]: Same, no opinion. We can be fast/loose w/ sample structure because we can change as needed.

File: docs/develop/python/message-passing.mdx
Comment (cretz) [reply docs/develop/python/message-passing.mdx:66]: Maybe a logger statement somewhere saying it was approved by someone, unsure. No big deal leaving unused field here though.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Right! Not sure why I was doing that. (Just `list(self.greetings)` is pretty idiomatic in Python.)

**Review by drewhoskins-temporal — APPROVED** [review]
great, thanks for taking the time to make this good!

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:33]: > I'm going to push for maintaining single-file structure for simple toy samples like this, so that they can simply be run by `python the_example.py` from within a virtualenv.

Actually, no, I have changed my mind. It's very hard for users to understand what's happening when there are errors if they are running a single-file sample with starter and worker stdin & stderr all going to a single terminal. I'm splitting it into `starter/worker/workflow` and am going to rename the directory from `updates_and_signals => message_passing`:

```
message_passing
├── introduction
└── safe_message_handlers
```

**Review by cretz — APPROVED** [review]
LGTM. Needs docs approval before merge of course.

File: docs/develop/python/message-passing.mdx
Comment (axfelix) [reply docs/develop/python/message-passing.mdx:35]: Thanks for calling this out -- indeed, we don't love having the huge sample at the top, it breaks with our docs conventions and I don't think it necessarily makes the page easier to read. In this case, I don't think the scaffolding adds very much, since it's relatively straightforward to explain Signals, Queries, and Updates individually -- no need to have them all in one block like that.

Comment (axfelix) [reply docs/develop/python/message-passing.mdx:35]: I know there was a desire to get docs' eyes on this ASAP so porting to other languages can begin, and I would say that this is the primary change I'd like to see before ports; other edits will be relatively minor.

File: docs/develop/python/message-passing.mdx
Comment (napcs) [inline docs/develop/python/message-passing.mdx:5]:
```suggestion
description: Develop with Queries, Signals, and Updates using the Temporal Python SDK.
```

Should consider putting back some of the meta description here though for SEO.

File: docs/develop/python/message-passing.mdx
Comment (napcs) [inline docs/develop/python/message-passing.mdx:35]: These code snippets should be pulled in from external code repositories throughout, not embedded directly.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Hi @axfelix, I've pushed a change that breaks apart the message handler code snippets and sends them to their respective sections. With the simpler structure and reduction in prose content it's still very easy to find the code that way, so I think both solutions have their merits, and don't dislike the broken out form. What we were investigating with the other form was creating a code sample that allows some developers to learn quickly from code, without navigating prose sections.

File: docs/develop/python/message-passing.mdx
Comment (cretz) [inline docs/develop/python/message-passing.mdx:349]: There is no example at the top of the page anymore

File: docs/develop/python/message-passing.mdx
Comment (axfelix) [reply docs/develop/python/message-passing.mdx:335]:
```suggestion
- [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) `"workflow execution already completed"`
```

File: docs/develop/python/message-passing.mdx
Comment (axfelix) [inline docs/develop/python/message-passing.mdx:353]: I would try to limit the use of parenthesis here, especially since a capitalized "I.e." looks a bit strange.

File: docs/develop/python/message-passing.mdx
Comment (axfelix) [inline docs/develop/python/message-passing.mdx:41]: May want to also mention that you can include your handlers anywhere in your Workflow definition since this isn't true of all SDKs.

File: docs/develop/python/message-passing.mdx
Comment (axfelix) [inline docs/develop/python/message-passing.mdx:153]: This is a style guide issue, imo -- rather than directing someone to "find out" and "look at" something else, this should be expressed more straightforwardly, e.g., "The argument type should match the corresponding handler method... for example, X, Y."

**Comment by axfelix** [comment]
I'd like to see the introduction to this page stick slightly closer to the other pages in our SDK docs as right now I think it contrasts too much. Otherwise, though, I like the emphasis on implementation details the increases toward the end of the page, I like what you're doing, and I think this is pretty close to being able to be merged. I left a few more comments, but I don't think this needs another round of feedback from me. Will let others finish weighing in.

**Comment by dandavison** [comment]
Thanks for reviewing @axfelix.

> I'd like to see the introduction to this page stick slightly closer to the other pages in our SDK docs as right now I think it contrasts too much. 

I agree; I noticed this morning that the intro needed a bit more. Sorry it's been a bit of an evolving target during review due to our product launch aims.

File: docs/develop/python/message-passing.mdx
Comment (cretz) [inline docs/develop/python/message-passing.mdx:507]:
```suggestion
A Signal can be made dynamic by adding `dynamic=True` to the `@workflow.signal` decorator.
```

(just saw when translating these docs)

File: docs/develop/python/message-passing.mdx
Comment (cretz) [inline docs/develop/python/message-passing.mdx:504]: This section only covers signal

File: docs/develop/python/message-passing.mdx
Comment (cretz) [inline docs/develop/python/message-passing.mdx:524]: These dynamic workflow/activity docs have nothing to do with message passing. They should be part of the "core application" part (which they are in some SDKs like .NET)

**Comment by fairlydurable** [comment]
@drewhoskins-temporal  Is this ready to transition from SDK-1486 to EDU-671?

**Comment by dandavison** [comment]
> @drewhoskins-temporal Is this ready to transition from SDK-1486 to EDU-671?

Not quite -- I'll let you know when I've addressed everything remaining that I know of from my side.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:5]: Thanks, done!

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:43]: Can you expand on that @axfelix? (In Python handlers must be methods on the Workflow class annotated with one of the decorators.)

Comment (axfelix) [reply docs/develop/python/message-passing.mdx:43]: Yes -- they have to be methods on the Workflow class, but they can be defined anywhere. Compared to eg Go, where iirc there's no such thing as a handler other than inline.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:160]: Thanks, I don't disagree with your edit. FWIW what I was going for here is the following: I was trying to write this from the point of view of someone writing client-side code who needs to switch attention to the Workflow definition (maybe using IDE hover-over / command-click) to look up the signature of the handler method. But I can see that it's either more conversational/informal thn we want, or not actually more helpful than the more direct phrasing, or both!

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx:43]: I've added this bullet right at the top:

```markdown
General guidelines:

- Message handlers are defined as methods on the Workflow class, using one of the three decorators: `@workflow.query`, `@workflow.signal`, and `@workflow.update`.
```

Does that cover it? I think once Python programmers understand that they are methods, we can leave it to them to think of the possibilities for how they can define them. I.e. ranging from beginners who'll only know how to do it exactly as in our examples, to people who might choose inheritance, or mixins, or to inject them via class decorators or metaclasses!

File: docs/develop/python/message-passing.mdx
Comment (drewhoskins-temporal) [inline docs/develop/python/message-passing.mdx:110]: This links to #sending-updates whereas you have #send-updates.

**Comment by dandavison** [comment]
> Is this ready to transition from SDK-1486 to EDU-671?

Yes, over to docs team!

**Comment by dandavison** [comment]
Erica's changes are merged into this one; this PR is now just waiting for an approval from docs team.

**Review by axfelix — APPROVED** [review]
Stamped! Thanks for working with us so much on this one. Really happy with how it turned out.

**Comment by dandavison** [comment]
Thanks @axfelix and @fairlydurable.

I've added two final commits for @fairlydurable to review:

1. changing a code snippet only to address an uresolved review comment above: https://github.com/temporalio/documentation/pull/2976/commits/2820990b7ebd45ccdb5b72dfe70349c286d0c655

2. Adding a line documenting `workflow.current_update_info` which should be included but was not: https://github.com/temporalio/documentation/commit/6446f01d78293aa42aa3be6766b8bc752ee1c11b 


I'll obviously let you know if I think of any more omissions, but other than that over to you for merge.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks, yes I emphasized that `wait_condition` is very useful in order to try to avoid it being overlooked.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks, that got done.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: [Agreed via other channels that we'd do the snipsync linkage in subsequent PRs]

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Good catch! (Pre-existing error in docs. d48ef208d780fa163eec6f022b03c13b2df11ae8)

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Yes, agreed. Opened https://temporalio.atlassian.net/browse/EDU-3058

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: It's not great, but I think it's OK to merge. It shows how to use it for signal, by way of an example. The whole "dynamic XXX" topic needs more work, and wasn't the focus of this PR.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks, that was changed and I think you'll be OK with the current text.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks, the "deliberate" language isn't there any longer.

File: docs/develop/python/message-passing.mdx
Comment (dandavison) [reply docs/develop/python/message-passing.mdx]: Thanks! (I actually moved to giving the RPC codes rather than the error messages)

**Pull request merged**
